pkg/controllers/webhookauthenticator/webhookauthenticator_controller.go

```diff
@@ -0,0 +1,254 @@
+package webhookauthenticator
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/informers"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	configinformers "github.com/openshift/client-go/config/informers/externalversions"
+	operatorconfigclient "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/controllers/common"
+)
+
+var knownConditionNames = sets.NewString(
+	"AuthenticatorAuthConfigDegraded",
+	"AuthenticatorKubeConfigSyncDegraded",
+	"AuthenticatorOAuthAPIServiceDegraded",
+	"AuthenticatorOperatorStatusDegraded",
+	"OpenShiftAuthenticationAvailable",
+)
+
+type webhookAuthenticatorController struct {
+	authentication  configv1client.AuthenticationInterface
+	serviceAccounts corev1client.ServiceAccountsGetter
+
+	saLister  corev1listers.ServiceAccountLister
+	svcLister corev1listers.ServiceLister
+	secrets   corev1client.SecretsGetter
+
+	operatorConfigClient operatorconfigclient.AuthenticationsGetter
+	operatorClient       v1helpers.OperatorClient
+}
+
+func NewWebhookAuthenticatorController(
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	configInformer configinformers.SharedInformerFactory,
+	secrets corev1client.SecretsGetter,
+	serviceAccounts corev1client.ServiceAccountsGetter,
+	authentication configv1client.AuthenticationInterface,
+	operatorConfigClient operatorconfigclient.AuthenticationsGetter,
+	operatorClient v1helpers.OperatorClient,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &webhookAuthenticatorController{
+		secrets:              secrets,
+		serviceAccounts:      serviceAccounts,
+		svcLister:            kubeInformersForTargetNamespace.Core().V1().Services().Lister(),
+		saLister:             kubeInformersForTargetNamespace.Core().V1().ServiceAccounts().Lister(),
+		authentication:       authentication,
+		operatorConfigClient: operatorConfigClient,
+		operatorClient:       operatorClient,
+	}
+	return factory.New().WithInformers(
```

Comment:
`WithSyncDegradedOnError` is your friend. makes the rest of this much easier to reason about and removes the need for "known conditons"